Add Creation parsing and fix issues

Silarn opened the pull request:
- Read creations from LocalAppData
- Don't include CC files in primary plugins
- Don't parse CCC file if sTestFile set
- Fix parsing of secondary data directories

TODO: Consolidate ContentCatalog.txt parsing somewhat?

Silarn committed:
Add Creation parsing and fix issues
- Read creations from LocalAppData
- Don't include CC files in primary plugins
- Don't parse CCC file if sTestFile set
- Fix parsing of secondary data directories

Silarn committed:
Consolidate parsing, read failure error check

Silarn committed:
Implement Starfield.ccc virtualization
- Fix issues with core plugin LO

Silarn committed:
Reorder primary plugins to better comport with base LO

Holt59 commented on src/gamestarfield.cpp:
```diff
@@ -52,12 +56,35 @@ bool GameStarfield::init(IOrganizer* moInfo)
+  if (m_Organizer->pluginSetting(name(), "enable_loadorder_fix").toBool()) {
+    QFile cccFile(m_Organizer->profilePath() + "/Starfield.ccc");
+    if (cccFile.open(QIODevice::WriteOnly)) {
+      ON_BLOCK_EXIT([&cccFile]() {
```
I don't think that block is required, the destructor of `QFile` will close the file.

Silarn replied:
Sure. In that case I should remove the example that already existed in the file.

Holt59 commented on src/gamestarfield.cpp:
```diff
@@ -272,29 +308,63 @@ QStringList GameStarfield::DLCPlugins() const
-      if ((line.size() > 0) && (line.at(0) != '#')) {
-        modName = QString::fromUtf8(line.constData()).toLower();
+    if (file.open(QIODevice::ReadOnly)) {
+      ON_BLOCK_EXIT([&file]() {
```
Same as above.

Holt59 reviewed (changes requested):
I do not own Starfield so I am not sure what the purpose of this is but I think 1) the parsing section of code should be moved to dedicated functions and 2) this should be more commented / documented than it is right now.

Silarn commented:
> I do not own Starfield so I am not sure what the purpose of this is but I think 1) the parsing section of code should be moved to dedicated functions and 2) this should be more commented / documented than it is right now.

There is no default CCC file in Starfield. ContentCatalog.txt is the only way to look up what mods constitute Creation mods. However the CCC file will still be parsed if it exists. Crucially, core plugins are not auto-loaded before plugins.txt. One workaround is to put the core plugins into a new Starfield.ccc file which forces the game to load them first.

So this function must parse both in order to determine what we're considering Creations. However we don't want it to duplicate the core plugins if they get put in there as this causes MO2 to bug out trying to load duplicate core plugins.

I guess I can add some comments but I don't really see the point of moving the parsing code out. This is the one place it needs to be, and pulling it from the unmanaged mods is a bit odd. I don't think it makes sense to make it part of the interfaces. Perhaps it could be recast as the Starfield class.

Silarn committed:
Remove scope guards, add comments

Silarn commented:
Actually I just realized there's an issue in the CCC parsing code. It should be adding the files if they aren't core plugins. I'm surprised that didn't trigger an issue, honestly. I'll fix that when I'm back at my computer.

Holt59 commented on src/starfieldunmanagedmods.cpp:
```diff
@@ -16,32 +25,121 @@ QStringList StarfieldUnmangedMods::mods(bool onlyOfficial) const
 
 QString StarfieldUnmangedMods::displayName(const QString& modName) const
 {
+  QFile content(m_AppDataFolder + "/" + game()->gameShortName() +
```
I think this declaration can be removed.

Holt59 commented on src/starfieldunmanagedmods.cpp:
```diff
@@ -16,32 +25,121 @@ QStringList StarfieldUnmangedMods::mods(bool onlyOfficial) const
+                "/ContentCatalog.txt");
+  QString name           = modName;
+  QJsonObject contentObj = getContentCatalog();
+  for (auto mod : contentObj.keys()) {
```
This should probably be `const auto&` (applies to most `for`-loop I think).

Holt59 commented on src/starfieldunmanagedmods.cpp:
```diff
@@ -16,32 +25,121 @@ QStringList StarfieldUnmangedMods::mods(bool onlyOfficial) const
+  for (auto mod : contentObj.keys()) {
+    if (mod == "ContentCatalog")
+      continue;
+    auto modData  = contentObj.value(mod).toObject();
```
You could directory iterate `contentObj` (not `keys`) to access both key and value.

Silarn replied:
This actually returns a QJsonValueRef which doesn't give you access to the key. You can use the QJsonObject::iterator with an old-style for loop. I don't necessarily see how that's any better than what I'm doing now.

Holt59 replied:
Is that true? I have not tested it but according to https://doc.qt.io/qt-6/qjsonobject.html, `begin()` and `end()` returns `QJsonObject::iterator` (https://doc.qt.io/qt-6/qjsonobject-iterator.html) with both a `.key()` and a `.value()` getters.

Silarn replied:
Yes, if you use the begin; != end; iterator++ method. A pure auto on just the object returns the valueref which doesn't give you the key.

Holt59 replied:
Ah yes you're right they did not implement proper structured binding...

Holt59 commented on src/starfieldunmanagedmods.h:
```diff
@@ -2,17 +2,27 @@
 {
 public:
-  StarfieldUnmangedMods(const GameGamebryo* game);
+  StarfieldUnmangedMods(const GameStarfield* game, const QString appDataFolder);
```
Missing  `&` for `appDataFolder`.

Holt59 commented on src/starfieldunmanagedmods.cpp:
```diff
@@ -16,32 +25,121 @@ QStringList StarfieldUnmangedMods::mods(bool onlyOfficial) const
 {
+  QStringList files;
+  QJsonObject contentObj = getContentCatalog();
+  for (auto mod : contentObj.keys()) {
```
Same comment as below: can iterate items directly and should use `const auto&`.

Holt59 commented on src/starfieldunmanagedmods.cpp:
```diff
@@ -16,32 +25,121 @@ QStringList StarfieldUnmangedMods::mods(bool onlyOfficial) const
+                "/ContentCatalog.txt");
+  if (content.exists()) {
+    if (content.open(QIODevice::OpenModeFlag::ReadOnly)) {
+      ON_BLOCK_EXIT([&content]() {
```
Can be removed.

Holt59 commented on src/starfieldunmanagedmods.cpp:
```diff
@@ -16,32 +25,121 @@ QStringList StarfieldUnmangedMods::mods(bool onlyOfficial) const
+{
+  QFile content(m_AppDataFolder + "/" + game()->gameShortName() +
+                "/ContentCatalog.txt");
+  if (content.exists()) {
```
I would reverse the `if` to avoid the three nested levels:

```cpp
if (!content.exists()) {
  return {};
}

if (!content.open(QIODevice::OpenModeFlag::ReadOnly)) {
  return {};
}

// ...
```

Holt59 commented on src/gamestarfield.cpp:
```diff
@@ -272,26 +311,59 @@ QStringList GameStarfield::DLCPlugins() const
+    for (auto mod : contentObj.keys()) {
+      if (mod == "ContentCatalog")
+        continue;
+      auto modData  = contentObj.value(mod).toObject();
```
After reading everything, I think this should be refactored somewhere so that `getContentCatalog` returns a custom structure with appropriate type rather that the `QJsonObject`. That would make using it in other functions more straightforward.

Silarn commented:
I'll make some of these changes when I can. I doubt I even need the exists checks as open will simply return false.

I suppose I can feed the data into a basic structure with just the files and mod name which would skip having to ignore the 'ContentCatalog' object and simplify loading the various mod data types. The main thing is restructuring it to use the plugin filename as the key.

Silarn committed:
Various changes and fixes
- Fix typo in unmanaged mods class name
- Add struct to store content catalog data
- Fetch relevant info in content catalog parser
- Use unmanaged mods parseer in main game class (for CC plugins)
- for loop fixes
